interfaces/apparmor/spec.go

```diff
@@ -173,6 +174,92 @@ func isProbablyPresent(path string) bool {
 	return path == "/" || path == "/snap" || path == "/var" || path == "/var/snap" || path == "/tmp" || path == "/usr" || path == "/etc"
 }
 
+// chopTree takes a path and depth and returns two lists of apparmor path expressions.
+//
+// The returned lists of expressions are referred to as left and right.
+//
+// The left list describes directories at depth up to and including
+// assumedPrefixDepth and can be used to grant read permission to them (by
+// appending the string "r, " to each element). This corresponds to an
+// assumption about those directories being present in the system and being
+// just traversed. Note that depth is defined as the number of directories
+// traversed, including the root directory.
+//
+// The right list describes the remaining directories and the leaf entry
+// (either file or directory) and is somewhat subtle. At each depth level the
+// expression describes all the files and directories at that level. Coupled
+// with the string "rw ," it can be used to create a rule that allows write
+// access to any file therein, but not deeper.
+//
+// For example, with path: "/foo/bar/froz/baz" and depth 3 the result is:
+// []string{"/", "/foo/", "/foo/bar/"} and []string{"/foo/*", "/foo/bar/*",
+// "/foo/bar/froz/*"}. Coupled with the aforementioned constants this
+// translates to the following apparmor rules:
+//
+//   / r,
+//   /foo/ r,
+//   /foo/bar/ r,
+//
+//   /foo/bar/* rw,
+//   /foo/bar/*/ rw,
+//   /foo/bar/froz/* rw,
+//   /foo/bar/froz/*/ rw,
+//
+// Those rules are useful for constructing the apparmor profile for a writable
+// mimic that needs to be present in a specific directory (e.g. in
+// /foo/bar/froz/baz) assuming that part of that directory already exists (e.g.
+// /foo/bar/) but may need to be created earlier (e.g. in /foo/bar/froz).
+//
+// The mimic works by mounting a tmpfs over the mimicked directory and then
+// re-creating empty files and directories as mount points for the subsequent
+// bind- mount operations to latch onto. This is why the right list of
+// expressions use * and */, this allows the expressions to capture files and
+// directories at a specific path.
```

Comment:
Not a blocker: you should have enough information to differentiate with the rule (ie, only use * with a file and only use */ with a dir). I'm not sure how much this would complicate the code. It also isn't terribly interesting to make that distinction in the snap-update-ns profile.

Reply:
Are you sure? I thought about this as well but the reason for * and */ together is that we need mimic the directory at that specific level. This implies that we need to create empty files and empty directories to re-create the directory as it was there. It doesn't matter if the original path was for a file or for a directory.

Reply:
My point was that looking at this in isolation, if `{p: "/foo/bar/froz/baz/", d: 3`, then the lhs is clearly `{"/", "/foo/", "/foo/bar/"}` but the rhs could be `{"/foo/bar/*", "/foo/bar/*/", "/foo/bar/froz/*/"}`. However, in typing that, I remembered that part of the mimicking behavior outside of chopTree will be to mimic the files in /foo/bar/froz/* *and* `/baz/` since `froz` is being mimicked. The fact that I forgot about that indicates that perhaps the comment should be more clear that the caller of chopTree() expects both the file and directory rules in place to do its job.